streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/query/RealtimeJoinBolt.java

```diff
@@ -0,0 +1,638 @@
+        String[] fieldNames = commaSeparatedKeys.split(",");
+
+        outputFields = new FieldSelector[fieldNames.length];
+        for (int i = 0; i < fieldNames.length; i++) {
+            outputFields[i] = new FieldSelector(fieldNames[i]);
+        }
+        return this;
+    }
+
+    /** Convenience method for Streamline that prefixes each keyname with 'streamline-event.'
+     *
+     * @param commaSeparatedKeys
+     * @return
+     */
+    public RealtimeJoinBolt streamlineSelect(String commaSeparatedKeys) {
+        String prefixedKeys = convertToStreamLineKeys(commaSeparatedKeys);
+        streamLineProjection = true;
+        return  select(prefixedKeys);
+    }
+
+    public RealtimeJoinBolt withOutputStream(String streamName) {
+        this.outputStream = streamName;
+        return this;
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        if (timeBasedRetention) {
+            expireAndAckTimedOutEntries(lookupBuffer);
+        }
+
+        String streamId = getStreamSelector(tuple);
+        if ( isLookupStream(streamId) ) {
+            String key = makeLookupTupleKey(tuple);
+            if(dropOlderDuplicates)
+                lookupBuffer.removeAll(key);
+            lookupBuffer.put(key, new TupleInfo(tuple) );
+
+            if(timeBasedRetention) {
+                timeTracker.add(System.currentTimeMillis());
+            } else {  // count based Rotation
+                if (lookupBuffer.size() > retentionCount) {
+                    TupleInfo expired = removeHead(lookupBuffer);
+                    if( (joinType==JoinType.RIGHT) || (joinType==JoinType.OUTER)  ) {
+                        emitUnMatchedTuples(expired);
+                    }
+                    collector.ack(expired.tuple);
+                }
+            }
+        } else if (isDataStream(streamId) ) {
+            List<TupleInfo> matches = matchWithLookupStream(tuple);
+            if (matches==null || matches.isEmpty() ) {  // no match
+                if (joinType==JoinType.LEFT ||  joinType==JoinType.OUTER ) {
+                    ArrayList<Object> outputTuple = doProjection(tuple, null);
+                    emit(outputTuple, tuple);
+                    return;
+                }
+                collector.ack(tuple);
+            }
+
+            for (TupleInfo lookupTuple : matches) { // match found
+                lookupTuple.unmatched = false;
+                ArrayList<Object> outputTuple = doProjection(lookupTuple.tuple, tuple);
+                emit(outputTuple, tuple, lookupTuple.tuple);
+            }
+            collector.ack(tuple);
+        } else {
+            LOG.warn("Received tuple from unexpected stream/source : {}. Tuple will be dropped.", streamId);
+        }
+    }
+
+    private String makeLookupTupleKey(Tuple tuple) {
+        StringBuffer key = new StringBuffer();
+        for (JoinInfo ji : joinCriteria) {
+            String partialKey = findField(ji.lookupField, tuple).toString();
```

Comment:
good catch. i have not handled the case of invalid tuples properly. will address this.